### PR TITLE
8269268: JDWP: Properly fix thread lookup assert in findThread()

### PR DESCRIPTION
**Pull request by plummercj**
Re-enable the assert that was disabled (with some overhead) by [JDK-8265683](https://bugs.openjdk.java.net/browse/JDK-8265683). Explanation is in the CR and also in comments included with the changes.

I tested by running `vmTestbase/nsk/jdb/suspend/suspend001/suspend001.java` and `vmTestbase/nsk/jdb/wherei/wherei001/wherei001.java` 100's of times, and did not see any failures. I also verified the original issue was still reproducible by temporarily not setting `gdata->handlingVMDeath = JNI_TRUE`, which did trigger the assert as expected.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269268](https://bugs.openjdk.java.net/browse/JDK-8269268): JDWP: Properly fix thread lookup assert in findThread()

### Reviewers
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer)
 * [Alex Menkov](https://openjdk.java.net/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4580/head:pull/4580` \
`$ git checkout pull/4580`

Update a local copy of the PR: \
`$ git checkout pull/4580` \
`$ git pull https://git.openjdk.java.net/jdk pull/4580/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4580`

View PR using the GUI difftool: \
`$ git pr show -t 4580`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4580.diff">https://git.openjdk.java.net/jdk/pull/4580.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c [hunk 1]
```diff
@@ -1239,6 +1239,10 @@ cbVMDeath(jvmtiEnv *jvmti_env, JNIEnv *env)
         EXIT_ERROR(error,"Can't clear event callbacks on vm death");
     }
 
+    /* Setting this flag is needed for a very special case. See the reference in findThread(). */
+    gdata->handlingVMDeath = JNI_TRUE;
```
Comment (dholmes-ora) [inline src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:1243]: Is this used to communicate across different threads? If so then it needs to be at least volatile, but a more forceful memory-barrier may be needed on some platforms. I can't tell in what context findThread will be called after this has been set to true.

Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:1243]: I had an answer typed up for you, but I realized I need to think some things through a bit more, and it's getting late. I'll look into it some more tomorrow.

**Comment by plummercj** [comment]
There are two threads to consider here. The one that `VM_DEATH` was received on and the thread that reads JDWP commands from the debugger. For the latter, a `VM.Resume` command has been received. The debugger does this as the test wraps up, allowing all debuggee threads to resume and the debugeee to exit. `VM.resume()` calls `threadControl_resumeAll()`, which calls `commonResumeList()`, which walks the `runningThreads` list to build an array of jthreads that need to be resumed. This array is passed to JVMTI `ResumeThreadList()`. Any time after this point a `VM_DEATH` can be triggered since the debuggee main thread calls `System.exit()` and does not wait for all the threads it creates to exit first (no join is done). The `VM_DEATH` event leads to clearing all the JVMTI callbacks, so the debug agent is no longer notified of `THREAD_END` events as the other threads created by the debuggee exit. This leads to threads still being on `runningThreads`, even though they have exited and had their TLS cleared by JVMTI. Keep in mind we are still in `commonResumeList()` when all this is happening. After calling JVMTI `ResumeThreadList()`, it then walks that same array of jthreads that were resumed, doing a `findThread()` on each. This is done mainly so it can update the `suspendCount` the `ThreadNodes`. This is when the assert was getting triggered. Some of the debuggee threads had already exited and had their TLS cleared, but were still in `runningThreads` since the `THREAD_END` was never received. The fix is to not assert (and instead look up the thread in `runningThreads`) if we are dealing with a `VM_DEATH` event.

You mentioned `gdata->handlingVMDeath` should be volatile. I will fix that. I also think I will rename it to `gdata->jvmtiCallBacksCleared`. There is no need to ever set it back false, so I will remove that line of code. Once true, we know the vm is exiting anyway, and the callbacks will never be setup again. Also, I'm going to set it true before actually clearing the callbacks rather than after. Doing it early is fine. In fact if it was always true, `findThread()` would still work, but just be slower.

As far as any other synchronizing or memory barriers being needed, I'm a bit unclear on that part. The only thing that matters is that `gdata->jvmtiCallBacksCleared` being set true is visible to other threads before the callbacks get cleared. It seems the call to `SetEventCallbacks()` should ensure this.

**Review by kevinjwalls — APPROVED** [review]

**Review by alexmenkov — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
LGTM.
I like the way it is fixed.
Thanks,
Serguei

**Comment by dholmes-ora** [comment]
Just to follow up:

> As far as any other synchronizing or memory barriers being needed, I'm a bit unclear on that part. The only thing that matters is that gdata->jvmtiCallBacksCleared being set true is visible to other threads before the callbacks get cleared. It seems the call to SetEventCallbacks() should ensure this.

In practice this should be fine as `SetEventCallbacks` acquires a monitor/mutex; and `findThread` is also called while holding a monitor, so although different monitors, in practice the acquire/release semantics should make this work. Additionally the complexity of the interaction between the two threads involved will introduce other synchronization actions.

**Pull request closed without merging**